src/controller/rpc.cpp

```diff
@@ -0,0 +1,208 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2020 - 2022 Pionix GmbH and Contributors to EVerest
+
+#include "rpc.hpp"
+
+#include "command_api.hpp"
+#include "ipc.hpp"
+
+enum class RPCRequestErrorCode
+{
+    ParseError = -32700,
+    InvalidRequest = -32600,
+    MethodNotFound = -32601,
+    InvalidParams = -32602,
+    InternalError = -32603
+};
+
+class RPCRequestError : public std::runtime_error {
+public:
+    RPCRequestError(RPCRequestErrorCode error_code, const std::string& message, const nlohmann::json& id) :
+        std::runtime_error(
+            "Error code: " + std::to_string(static_cast<std::underlying_type_t<RPCRequestErrorCode>>(error_code)) +
+            "\n" + message),
+        error_code(error_code),
+        message(message),
+        id(id) {
+    }
+
+    RPCRequestErrorCode get_error_code() const {
+        return error_code;
+    }
+
+    const std::string& get_message() const {
+        return this->message;
+    }
+
+    const nlohmann::json& get_id() const {
+        return this->id;
+    }
+
+private:
+    RPCRequestErrorCode error_code;
+    std::string message;
+    nlohmann::json id;
+};
+
+class RPCRequest {
```

Comment:
What's the reason for defining this class directly in the rpc.cpp not in the rcp.hpp ?